src/Lib.P2P/Multiplex/Substream.cs

```diff
@@ -0,0 +1,240 @@
+
+        /// <inheritdoc />
+        public override long Position
+        {
+            get => throw new NotSupportedException();
+            set => throw new NotSupportedException();
+        }
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin) { throw new NotSupportedException(); }
+
+        /// <inheritdoc />
+        public override void SetLength(long value) { throw new NotSupportedException(); }
+
+        /// <summary>
+        ///   Add some data that should be read by the stream.
+        /// </summary>
+        /// <param name="data">
+        ///   The data to be read.
+        /// </param>
+        /// <remarks>
+        ///   <b>AddData</b> is called when the muxer receives a packet for this
+        ///   stream.
+        /// </remarks>
+        public void AddData(byte[] data) { _inBlocks.Post(data); }
+
+        /// <summary>
+        ///   Indicates that the stream will not receive any more data.
+        /// </summary>
+        /// <seealso cref="AddData(byte[])"/>
+        /// <remarks>
+        ///   <b>NoMoreData</b> is called when the muxer receives a packet to
+        ///   close this stream.
+        /// </remarks>
+        public void NoMoreData() { _inBlocks.Complete(); }
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+#pragma warning disable VSTHRD002 
+            return ReadAsync(buffer, offset, count).GetAwaiter().GetResult();
+#pragma warning restore VSTHRD002 
+        }
+
+        /// <inheritdoc />
+        public override async Task<int> ReadAsync(byte[] buffer,
+            int offset,
+            int count,
+            CancellationToken cancellationToken)
+        {
+            var total = 0;
+            while (count > 0 && !_eos)
+
+                // Does the current block have some unread data?
+                if (_inBlock != null && _inBlockOffset < _inBlock.Length)
+                {
+                    var n = Math.Min(_inBlock.Length - _inBlockOffset, count);
+                    Array.Copy(_inBlock, _inBlockOffset, buffer, offset, n);
+                    total += n;
+                    count -= n;
+                    offset += n;
+                    _inBlockOffset += n;
+                }
+
+                // Otherwise, wait for a new block of data.
+                else
+                {
+                    try
+                    {
+                        _inBlock = await _inBlocks.ReceiveAsync(cancellationToken).ConfigureAwait(false);
+                        _inBlockOffset = 0;
+                    }
+                    catch (InvalidOperationException) // no more data!
+                    {
+                        _eos = true;
+                    }
+                }
+
+            return total;
+        }
+
+        /// <inheritdoc />
+        public override void Flush()
+        {
+#pragma warning disable VSTHRD002 
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Do not suppress issues.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)